<a href="https://colab.research.google.com/github/js-valencia/Algebra/blob/main/Algebra_grupal.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Create a Python code to solve a 3x3 system of linear equations using Cramer's rule with cofactors. The code should print the step-by-step process, the final solution, and include the user's personal information.

## Definir el sistema de ecuaciones

### Subtask:
Ingresar los coeficientes del sistema de ecuaciones 3x3.


**Reasoning**:
Prompt the user for the coefficients of the 3x3 system of equations and their personal information.



In [8]:
print("Introduce los coeficientes del sistema de ecuaciones lineales 3x3. (ax + by + cz = d):")

a1 = float(input("Ingresar a1: "))
b1 = float(input("Ingresar  b1: "))
c1 = float(input("Ingresar  c1: "))
d1 = float(input("Ingresar  d1: "))

a2 = float(input("Ingresar  a2: "))
b2 = float(input("Ingresar  b2: "))
c2 = float(input("Ingresar  c2: "))
d2 = float(input("Ingresar  d2: "))

a3 = float(input("Ingresar  a3: "))
b3 = float(input("Ingresar  b3: "))
c3 = float(input("Ingresar  c3: "))
d3 = float(input("Ingresar  d3: "))

print("\nDatos Informativos:")
name1 = input("Ingrese el nombre del primer integrante: ")
name2 = input("Ingrese el nombre del segundo integrante: ")
id_number = input("Ingrese su carrera: ")
fecha = input("Ingrese la fecha: ")

Introduce los coeficientes del sistema de ecuaciones lineales 3x3. (ax + by + cz = d):
Ingresar a1: 120
Ingresar  b1: 80
Ingresar  c1: 100
Ingresar  d1: 1540
Ingresar  a2: 200
Ingresar  b2: 100
Ingresar  c2: 150
Ingresar  d2: 2620
Ingresar  a3: 150
Ingresar  b3: 60
Ingresar  c3: 200
Ingresar  d3: 2290

Datos Informativos:
Ingrese el nombre del primer integrante: José Valencia
Ingrese el nombre del segundo integrante: Javier Guarnizo
Ingrese su carrera: Computación
Ingrese la fecha: 28/10/2025


## Calcular el determinante principal

### Subtask:
Calcular el determinante de la matriz de coeficientes.


**Reasoning**:
Define a function to calculate the determinant of a 3x3 matrix using the cofactor expansion method, call the function with the main coefficients, store the result, and print it.



In [9]:
def determinant_3x3(a1, b1, c1, a2, b2, c2, a3, b3, c3):
  """Calculates the determinant of a 3x3 matrix using cofactor expansion."""
  det = a1 * (b2 * c3 - b3 * c2) - b1 * (a2 * c3 - a3 * c2) + c1 * (a2 * b3 - a3 * b2)
  return det

det_main = determinant_3x3(a1, b1, c1, a2, b2, c2, a3, b3, c3)
print("Determinante principal:", det_main)

Determinante principal: -380000.0


## Calcular los determinantes para cada variable

### Subtask:
Para cada variable (x, y, z), reemplazar la columna de coeficientes correspondiente con la columna de términos independientes y calcular el determinante de la nueva matriz.


**Reasoning**:
Define the `determinant_3x3` function and calculate the determinants for x, y, and z by replacing the respective columns with the independent terms and using the defined function. Finally, print these determinants.



In [3]:
def determinant_3x3(a1, b1, c1, a2, b2, c2, a3, b3, c3):
  """Calculates the determinant of a 3x3 matrix using cofactor expansion."""
  det = a1 * (b2 * c3 - b3 * c2) - b1 * (a2 * c3 - a3 * c2) + c1 * (a2 * b3 - a3 * b2)
  return det

det_x = determinant_3x3(d1, b1, c1, d2, b2, c2, d3, b3, c3)
det_y = determinant_3x3(a1, d1, c1, a2, d2, c2, a3, d3, c3)
det_z = determinant_3x3(a1, b1, d1, a2, b2, d2, a3, b3, d3)

print("Determinante para x:", det_x)
print("Determinante para y:", det_y)
print("Determinante para z:", det_z)

Determinante para x: -4680000.0
Determinante para y: 1210000.0
Determinante para z: -1204000.0


## Resolver el sistema

### Subtask:
Dividir el determinante de cada variable por el determinante principal para encontrar el valor de cada variable.


**Reasoning**:
Check if the main determinant is zero and calculate the values of x, y, and z if it is not.



In [11]:
if det_main == 0:
    print("El sistema no tiene solución única y no puede resolverse mediante la regla de Cramer.")
else:
    x = det_x / det_main
    y = det_y / det_main
    z = det_z / det_main
    print("La solución es:")
    print("x =", x)
    print("y =", y)
    print("z =", z)

La solución es:
x = 12.31578947368421
y = -3.1842105263157894
z = 3.168421052631579


## Mostrar el proceso y la respuesta

### Subtask:
Imprimir los determinantes calculados, los pasos de la resolución y los valores de las variables.


**Reasoning**:
Print the calculated determinants, the steps of the solution using Cramer's rule, and the final values of the variables.



In [12]:
print("\n--- Detalles de la Solución ---")

# Step 1: Print the calculated determinants
print(f"\nDeterminantes Calculadas:")
print(f"Determinante del coeficiente de la matríz principal (det_main): {det_main}")
print(f"Determinante para x (det_x): {det_x}")
print(f"Determinante para y (det_y): {det_y}")
print(f"Determinante para z (det_z): {det_z}")

# Step 2: Print a step-by-step explanation of the solution using Cramer's rule
print("\nUsando al regla de Cramer paso por paso:")
print("La regla de Cramer establece que, para un sistema de ecuaciones lineales, el valor de cada variable se obtiene dividiendo el determinante de la matriz donde la columna de coeficientes de dicha variable se reemplaza por los términos constantes (d1, d2, d3) entre el determinante de la matriz de coeficientes principal.")

if det_main == 0:
    print("\nDado que el determinante principal es 0, el sistema no tiene solución única y no se puede aplicar la regla de Cramer.")
else:
    print(f"\nPara encontrar x: x = det_x / det_main = {det_x} / {det_main}")
    print(f"Para encontrar y: y = det_y / det_main = {det_y} / {det_main}")
    print(f"Para encontrar z: z = det_z / det_main = {det_z} / {det_main}")

# Step 3: Print the final calculated values of x, y, and z
print("\nSolución finaln:")
print(f"x = {x}")
print(f"y = {y}")
print(f"z = {z}")

print(f"\nSolucion para {name1, name2}:")
print(f"x = {x}")
print(f"y = {y}")
print(f"z = {z}")


--- Detalles de la Solución ---

Determinantes Calculadas:
Determinante del coeficiente de la matríz principal (det_main): -380000.0
Determinante para x (det_x): -4680000.0
Determinante para y (det_y): 1210000.0
Determinante para z (det_z): -1204000.0

Usando al regla de Cramer paso por paso:
La regla de Cramer establece que, para un sistema de ecuaciones lineales, el valor de cada variable se obtiene dividiendo el determinante de la matriz donde la columna de coeficientes de dicha variable se reemplaza por los términos constantes (d1, d2, d3) entre el determinante de la matriz de coeficientes principal.

Para encontrar x: x = det_x / det_main = -4680000.0 / -380000.0
Para encontrar y: y = det_y / det_main = 1210000.0 / -380000.0
Para encontrar z: z = det_z / det_main = -1204000.0 / -380000.0

Solución finaln:
x = 12.31578947368421
y = -3.1842105263157894
z = 3.168421052631579

Solucion para ('José Valencia', 'Javier Guarnizo'):
x = 12.31578947368421
y = -3.1842105263157894
z = 3.1684

## Incluir datos informativos

### Subtask:
Agregar la información personal del usuario al final de la salida.


**Reasoning**:
Print the personal information of the user.



In [13]:
print("\n--- Datos Informativos: ---")
print("Integrantes: ", name1, name2)
print("Carrera: ", id_number)
print("Fecha: ", fecha)


--- Datos Informativos: ---
Integrantes:  José Valencia Javier Guarnizo
Carrera:  Computación
Fecha:  28/10/2025


## Summary:

### Data Analysis Key Findings

*   The system successfully collected the coefficients for a 3x3 system of linear equations and the user's personal information (name and ID).
*   The determinant of the main coefficient matrix (\-380000.0) was calculated using cofactor expansion.
*   The determinants for each variable (x, y, and z) were calculated by replacing the corresponding coefficient column with the independent terms. The calculated determinants were \-4680000.0 for x, 1210000.0 for y, and \-1204000.0 for z.
*   Since the main determinant was non-zero, the system has a unique solution.
*   The values for x, y, and z were calculated by dividing the respective variable determinants by the main determinant, resulting in x = 12.31578947368421, y = \-3.1842105263157894, and z = 3.168421052631579.
*   The code successfully printed the calculated determinants, a step-by-step explanation of Cramer's rule, the final solution for x, y, and z, and the user's personal information.

### Insights or Next Steps

*   The code provides a clear and effective step-by-step demonstration of solving a 3x3 system of linear equations using Cramer's rule.
*   Consider adding error handling for non-numeric input and potentially for systems with no unique solution beyond just detecting a zero main determinant (e.g., infinite solutions).
